### PR TITLE
Remove metrics nil checks and return the noop values from metrics package

### PR DESCRIPTION
**Pull request by madhuravi**

### CODE REVIEW

File: metrics/metrics.go [hunk 1]
```diff
@@ -98,5 +98,5 @@ func RootScope(i ScopeInit) (tally.Scope, tally.StatsReporter, io.Closer) {
 		}
 		return scope, reporter, closer
 	}
-	return nil, nil, nil
+	return tally.NoopScope, tally.NullStatsReporter, nil
```
Comment (sectioneight) [inline metrics/metrics.go]: Can we return a noop closer as well so people don't have to nil-check this?

Comment (madhuravi) [inline metrics/metrics.go]: Good point. However there is no default noop closer and tally does not provide a noop closer. I have one in testutils that I created for unit tests.
It seems contrived to create a Noop closer for this purpose in uberfx. We return nil in a lot of places and don't create noop object everytime.

**Comment by sectioneight** [comment]
I meant literally return https://golang.org/pkg/io/ioutil/#NopCloser passing
in `nil`. That way you can call Close() without a potential nil-deref

**Comment by madhuravi** [comment]
@sectioneight The signature for NopCloser is
func NopCloser(r io.Reader) io.ReadCloser

It takes in a reader. I think it would be confusing to construct a reader here and pass it in just to be able to use this.

**Comment by sectioneight** [comment]
A ReadCloser is a superset of a Closer. In other words, anything that takes an `io.Closer`, you can pass `ioutil.NopCloser(nil)`:

https://play.golang.org/p/GqeXIzTZdE

**Comment by madhuravi** [comment]
I don't mind use that. As a reader of this code I still think it's confusing, looking at metrics
func RootScope(i ScopeInit) (tally.Scope, tally.StatsReporter, io.Closer) {
  ...
  return tally.NoopScope, tally.NullStatsReporter, ioutil.NopCloser(nil)
}

What is ioutil doing here, pop over the method and you'll see "NopCloser returns a ReadCloser with a no-op Close method wrapping the provided Reader r." And this really has nothing to do with metrics. Seems nicer to have readability over an extra nil check.

**Comment by sectioneight** [comment]
As producers of APIs we should be providing ones that are as friendly to use as possible. Particularly when returning interfaces in a non-error case, it is extremely awkward to have to check if the interface return is `nil` or not, particularly due to the semantics of nil interfaces in Go. You don't have to use `ioutil.NopCloser` if you don't like how it reads, but having something that you can call `Close()` on (just like you have two other interfaces you can call methods on that do nothing) is much friendlier to developers and reads much more naturally than an extra `nil`.

File: metrics/metrics_test.go [hunk 2]
```diff
@@ -36,9 +37,9 @@ func TestRegisterReporter_OK(t *testing.T) {
 	defer cleanup()
 
 	scope, reporter, closer := getScope()
-	assert.Nil(t, scope)
-	assert.Nil(t, reporter)
-	assert.Nil(t, closer)
+	assert.Equal(t, scope, tally.NoopScope)
+	assert.Equal(t, reporter, tally.NullStatsReporter)
+	assert.Equal(t, ioutil.NopCloser(nil), closer)
```
Comment (sectioneight) [inline metrics/metrics_test.go]: This is an implementation detail. I'd do this instead:

```
assert.NoError(t, closer.Close())
```

**Pull request merged**